File: typescript/day_5/part_2.ts
import { readFileSync } from 'fs';

function solve(input: string): number {
    const sections = input.trim().split('\n\n');

    const ranges: [number, number][] = [];
    for (const line of sections[0].trim().split('\n')) {
        if (line.trim()) {
            const parts = line.split('-');
            const start = parseInt(parts[0]);
            const end = parseInt(parts[1]);
            ranges.push([start, end]);
        }
    }

    ranges.sort((a, b) => a[0] - b[0]);

    const merged: [number, number][] = [];
    for (const [start, end] of ranges) {
        if (merged.length > 0 && start <= merged[merged.length - 1][1] + 1) {
            merged[merged.length - 1][1] = Math.max(merged[merged.length - 1][1], end);
        } else {
            merged.push([start, end]);
        }
    }

    let total = 0;
    for (const [start, end] of merged) {
        const count = end - start + 1;
        total += count;
    }

    return total;
}

const input = readFileSync('../inputs/day_5.txt', 'utf-8');
const result = solve(input);
console.log(result);

File: java/day_5/Part2.java
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class Part2 {
    public static int solve(String input) {
        String[] sections = input.trim().split("\n\n");

        List<int[]> ranges = new ArrayList<>();
        for (String line : sections[0].trim().split("\n")) {
            if (!line.trim().isEmpty()) {
                String[] parts = line.split("-");
                int start = Integer.parseInt(parts[0]);
                int end = Integer.parseInt(parts[1]);
                ranges.add(new int[]{start, end});
            }
        }

        Collections.sort(ranges, (a, b) -> Integer.compare(a[0], b[0]));

        List<int[]> merged = new ArrayList<>();
        for (int[] r : ranges) {
            int start = r[0];
            int end = r[1];
            if (!merged.isEmpty() && start <= merged.get(merged.size() - 1)[1] + 1) {
                int[] last = merged.get(merged.size() - 1);
                last[1] = Math.max(last[1], end);
            } else {
                merged.add(new int[]{start, end});
            }
        }

        int total = 0;
        for (int[] r : merged) {
            int start = r[0];
            int end = r[1];
            int count = end - start + 1;
            total += count;
        }

        return total;
    }

    public static void main(String[] args) throws IOException {
        String input = new String(Files.readAllBytes(Paths.get("../inputs/day_5.txt")));
        int result = solve(input);
        System.out.println(result);
    }
}
